### PR TITLE
Remove calls to pkg/exit functions

### PR DESCRIPTION
**Pull request by qbarrand**
This PR removes calls to `pkg/exit` functions in `pkg` subpackages, which make the operator exit on errors.  
It is more efficient to return errors to the calling function, so that they can be properly processed and acted upon.

### CODE REVIEW

File: pkg/upgrade/upgrade.go [hunk 1]
```diff
@@ -151,27 +153,39 @@ func DriverToolkitVersion(entries []string, info map[string]NodeVersion) (map[st
...
 
 		if imageURL == "" {
 			warn.OnError(errors.New("No DTK image found, DTK cannot be used in a Build"))
 			return info, nil
 		}
 
-		if layer = registry.LastLayer(imageURL); layer == nil {
-			exit.OnError(errors.New("Cannot extract last layer for DTK from: " + imageURL))
+		if layer, err = registry.LastLayer(imageURL); layer == nil {
+			return nil, fmt.Errorf("cannot extract last layer for DTK from %s [%w]", imageURL, err)
```
Comment (pacevedom) [inline pkg/upgrade/upgrade.go]: Maybe use the same format as the rest of wrapping? `some text wrapper: original error`

File: pkg/resource/resource.go [hunk 2]
```diff
@@ -539,5 +564,5 @@ func checkForImagePullBackOff(obj *unstructured.Unstructured, namespace string)
 		return nil
 	}
 
-	return errors.New("Unexpected Phase of Pods in DameonSet: " + obj.GetName())
+	return fmt.Errorf("Unexpected Phase of Pods in DameonSet: %s ", obj.GetName())
```
Comment (pacevedom) [inline pkg/resource/resource.go]: White space at the end?

File: pkg/resource/resource.go [hunk 3]
```diff
@@ -280,41 +295,47 @@ func CRUD(obj *unstructured.Unstructured, releaseInstalled bool, owner v1.Object
...
+			return fmt.Errorf("could not set the owner reference: %w", err)
+		}
 
 		SetMetaData(obj, name, namespace)
 
-		if err := clients.Interface.Create(context.TODO(), obj); err != nil {
+		if err = clients.Interface.Create(context.TODO(), obj); err != nil {
 			if apierrors.IsForbidden(err) {
-				return errors.Wrap(err, "API error is forbidden")
+				return fmt.Errorf("API error: forbidden [%w]", err)
```
Comment (pacevedom) [inline pkg/resource/resource.go]: Maybe use the same format as other wrapped errors? `wrapper text: original error`

File: pkg/resource/resource.go [hunk 4]
```diff
@@ -94,22 +92,26 @@ func UpdateResourceVersion(req *unstructured.Unstructured, found *unstructured.U
...
 		clusterIP, fnd, err := unstructured.NestedString(found.Object, "spec", "clusterIP")
-		exit.OnErrorOrNotFound(fnd, err)
+		if err != nil || !fnd {
+			return fmt.Errorf("error or not found: %w", err)
+		}
 
-		if err := unstructured.SetNestedField(req.Object, clusterIP, "spec", "clusterIP"); err != nil {
-			return errors.Wrap(err, "Couldn't update clusterIP")
+		if err = unstructured.SetNestedField(req.Object, clusterIP, "spec", "clusterIP"); err != nil {
+			return fmt.Errorf("Couldn't update clusterIP: %w", err)
```
Comment (pacevedom) [inline pkg/resource/resource.go]: Wrapped error starts with capital letter.

File: pkg/resource/resource.go [hunk 5]
```diff
@@ -180,8 +184,9 @@ func CreateFromYAML(yamlFile []byte,
 			return errors.Wrap(err, "Could not convert yaml file to json"+string(yamlSpec))
 		}
 
-		err = obj.UnmarshalJSON(jsonSpec)
-		exit.OnError(errors.Wrap(err, "Cannot unmarshall json spec, check your manifest: "+string(jsonSpec)))
+		if err = obj.UnmarshalJSON(jsonSpec); err != nil {
+			return fmt.Errorf("cannot unmarshall json spec, check your manifest: %s [%w]", jsonSpec, err)
```
Comment (pacevedom) [inline pkg/resource/resource.go]: `wrapper text: original error`

File: pkg/proxy/proxy.go [hunk 6]
```diff
@@ -66,7 +68,9 @@ func SetupDaemonSet(obj *unstructured.Unstructured) error {
 func SetupPod(obj *unstructured.Unstructured) error {
 
 	containers, found, err := unstructured.NestedSlice(obj.Object, "spec", "containers")
-	exit.OnErrorOrNotFound(found, err)
+	if err != nil || !found {
+		return err
```
Comment (pacevedom) [inline pkg/proxy/proxy.go:76]: Distinguish between error and not found? If its not found err is nil and this is returned as if it worked ok.

File: pkg/proxy/proxy.go [hunk 7]
```diff
@@ -54,7 +54,9 @@ func Setup(obj *unstructured.Unstructured) error {
 // path... -> Pod, DaemonSet, BuildConfig, etc.
 func SetupDaemonSet(obj *unstructured.Unstructured) error {
 	containers, found, err := unstructured.NestedSlice(obj.Object, "spec", "template", "spec", "containers")
-	exit.OnErrorOrNotFound(found, err)
+	if err != nil || !found {
+		return err
```
Comment (pacevedom) [inline pkg/proxy/proxy.go:59]: Distinguish between error and not found? If its not found err is nil and this is returned as if it worked ok.

File: pkg/poll/poll.go [hunk 8]
```diff
@@ -504,7 +497,14 @@ func ForDaemonSetLogs(obj *unstructured.Unstructured, pattern string) error {
...
-		if match, _ := regexp.MatchString(pattern, lastBytes); !match {
+		var match bool
+
+		match, err = regexp.MatchString(pattern, lastBytes)
+		if err != nil {
+			return fmt.Errorf("error matching pattern %q: %v", pattern, err)
+		}
+
+		if !match {
 			return errors.New("Not yet done. Not matched against: " + pattern)
```
Comment (pacevedom) [inline pkg/poll/poll.go]: Use `fmt.Errorf`?

**Review by pacevedom — COMMENTED** [review]
Few nits

File: pkg/upgrade/upgrade.go (hunk 1 above)
Comment (qbarrand) [reply pkg/upgrade/upgrade.go]: Right, back to the regular format.

File: pkg/resource/resource.go (hunk 2 above)
Comment (qbarrand) [reply pkg/resource/resource.go]: Typo.

File: pkg/resource/resource.go (hunk 3 above)
Comment (qbarrand) [reply pkg/resource/resource.go]: Right, back to the regular format.

File: pkg/resource/resource.go (hunk 4 above)
Comment (qbarrand) [reply pkg/resource/resource.go]: Fixed.

File: pkg/resource/resource.go (hunk 5 above)
Comment (qbarrand) [reply pkg/resource/resource.go]: Right, back to the regular format.

File: pkg/proxy/proxy.go (hunk 6 above)
Comment (qbarrand) [reply pkg/proxy/proxy.go:76]: Oh indeed, well spotted.

File: pkg/proxy/proxy.go (hunk 7 above)
Comment (qbarrand) [reply pkg/proxy/proxy.go:59]: Fixed.

File: pkg/poll/poll.go (hunk 8 above)
Comment (qbarrand) [reply pkg/poll/poll.go]: Yes, done.

**Comment by ybettan** [comment]
We should probably `squash` all the commits here.
@qbarrand

**Comment by qbarrand** [comment]
@ybettan thanks for having a look. In fact, I forgot to mark this one as WIP. While working on fixing the CI, I tried applying all my changes into this PR to understand if removing `pkg/exit` entirely would fix the CI problem. It did not, so I will be splitting this PR into several smaller ones.

File: pkg/upgrade/upgrade.go [hunk 9]
```diff
@@ -151,27 +153,39 @@ func DriverToolkitVersion(entries []string, info map[string]NodeVersion) (map[st
...
+			return nil, err
+		}
+
+		if layer == nil {
 			continue
 		}
 		// For each entry we're fetching the cluster version and dtk URL
-		version, imageURL := registry.ReleaseManifests(layer)
+		version, imageURL, err := registry.ReleaseManifests(layer)
 		if version == "" {
```
Comment (pacevedom) [inline pkg/upgrade/upgrade.go]: I know the version and image will be empty if there is an error, but wouldnt it be more clear if we check on `err != nil` instead?

**Comment by qbarrand** [comment]
@pacevedom @ybettan PTAL

File: pkg/upgrade/upgrade.go (hunk 9 above)
Comment (qbarrand) [reply pkg/upgrade/upgrade.go]: Well spotted, thanks. Fixed and rebased.

**Comment by ybettan** [comment]
Thanks for squashing :) 

What is the purpose of that PR? why do we need to remove those calls? Maybe worth adding it in the commit message and in the PR description.

**Comment by qbarrand** [comment]
@ybettan description + commit message fixed 😄

**Pull request merged**